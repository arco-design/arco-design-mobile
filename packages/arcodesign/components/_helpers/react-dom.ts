import { ReactElement } from 'react';
import * as ReactDOM from 'react-dom';

function isObject(obj: any): obj is { [key: string]: any } {
    return Object.prototype.toString.call(obj) === '[object Object]';
}

export interface RootType {
    render: (element: ReactElement) => void;
    _unmount: () => void;
}
export interface RootTypeReact extends RootType {
    unmount?: () => void;
}
export type CreateRootFnType = (container: Element | DocumentFragment) => {
    render: (element: ReactElement) => void;
    unmount?: () => void;
};

// Cast ReactDOM to a version that might have createRoot and other properties
const typedReactDOM = ReactDOM as typeof ReactDOM & {
    createRoot?: CreateRootFnType;
    render?: (
        element: ReactElement,
        container: Element | DocumentFragment,
        callback?: () => void,
    ) => any;
    unmountComponentAtNode?: (container: Element | DocumentFragment) => boolean;
    __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED?: {
        usingClientEntryPoint?: boolean;
    };
};

let copyRender: (
    app: ReactElement,
    container: Element | DocumentFragment,
    createRootFunction?: CreateRootFnType,
) => RootType;

const updateUsingClientEntryPoint = (skipWarning?: boolean) => {
    // https://github.com/facebook/react/blob/17806594cc28284fe195f918e8d77de3516848ec/packages/react-dom/npm/client.js#L10
    // Avoid console warning
    const secretInternals = typedReactDOM.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    if (isObject(secretInternals)) {
        secretInternals.usingClientEntryPoint = skipWarning;
    }
};

const createRootFn = typedReactDOM.createRoot;

const getRender =
    (createRootFunction: CreateRootFnType) =>
    (app: ReactElement, container: Element | DocumentFragment) => {
        updateUsingClientEntryPoint(true);
        const root = createRootFunction!(container);
        updateUsingClientEntryPoint(false);

        root.render(app);

        return {
            render: (elementToRender: ReactElement) => {
                root.render(elementToRender);
            },
            _unmount() {
                setTimeout(() => {
                    if (root && typeof root.unmount === 'function') {
                        root.unmount();
                    }
                });
            },
        };
    };

if (createRootFn) {
    // React 18
    copyRender = getRender(createRootFn);
} else if (
    typeof typedReactDOM.render === 'function' &&
    typeof typedReactDOM.unmountComponentAtNode === 'function'
) {
    // React 16/17
    copyRender = function (app: ReactElement, container: Element | DocumentFragment) {
        typedReactDOM.render!(app, container); // Use non-null assertion
        return {
            render: (elementToRender: ReactElement) => {
                typedReactDOM.render!(elementToRender, container); // Use non-null assertion
            },
            _unmount() {
                typedReactDOM.unmountComponentAtNode!(container); // Use non-null assertion
            },
        };
    };
} else {
    copyRender = (
        app: ReactElement,
        container: Element | DocumentFragment,
        createRootFunction?: CreateRootFnType,
    ) => {
        const defaultCb = () => {
            // Fallback if no rendering method is found
            console.error(
                'ArcoDesign: ReactDOM.createRoot() or ReactDOM.render() not found. ' +
                    'This usually means you are using an unsupported version of React, ' +
                    'or ReactDOM is not properly initialized. ' +
                    'ArcoDesign Mobile React requires React 16, 17, 18, or 19.',
            );
            return { render: (_element: ReactElement) => {}, _unmount: () => {} };
        };
        if (createRootFunction) {
            return getRender(createRootFunction)(app, container);
        }
        return defaultCb();
    };
}

export const render = copyRender;
