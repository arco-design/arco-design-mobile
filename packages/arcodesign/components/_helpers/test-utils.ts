import React, { useEffect, useRef } from 'react';
import { Promise as ES6Promise, Promise } from 'es6-promise';

/**
 * React 19 兼容的 act 包装器
 * 用于包装可能触发状态更新的异步操作
 */
export const actWrapper = (callback: () => void | ES6Promise<void>): ES6Promise<void> => {
    return new ES6Promise<void>((resolve, reject) => {
        try {
            // 尝试使用 react-dom/test-utils 的 act
            const testUtils = require('react-dom/test-utils');
            if (testUtils && typeof testUtils.act === 'function') {
                // 对于同步回调
                if (typeof callback === 'function') {
                    try {
                        const result = callback();
                        if (result && typeof result.then === 'function') {
                            testUtils
                                .act(() => result)
                                .then(() => resolve())
                                .catch(reject);
                        } else {
                            testUtils.act(() => {
                                callback();
                            });
                            resolve();
                        }
                    } catch (error) {
                        reject(error);
                    }
                } else {
                    resolve();
                }
            } else {
                const result = callback();
                if (result && typeof result.then === 'function') {
                    result.then(() => resolve()).catch(reject);
                } else {
                    resolve();
                }
            }
        } catch (error) {
            // 如果 act 不可用，直接执行回调
            try {
                const result = callback();
                if (result && typeof result.then === 'function') {
                    result.then(() => resolve()).catch(reject);
                } else {
                    resolve();
                }
            } catch (callbackError) {
                reject(callbackError);
            }
        }
    });
};

/**
 * 清理未处理的异步操作，防止内存泄漏警告
 */
export const cleanupAsync = (): ES6Promise<void> => {
    return new ES6Promise<void>(resolve => {
        setTimeout(() => resolve(), 0);
    });
};

/**
 * 安全的状态更新，检查组件是否已卸载
 */
export const createSafeStateUpdater = (isMountedRef: React.MutableRefObject<boolean>) => {
    return (updater: () => void) => {
        if (isMountedRef.current) {
            updater();
        }
    };
};

/**
 * 用于测试的 useEffect 清理函数
 */
export const useTestCleanup = (cleanupFn: () => void) => {
    const isMountedRef = useRef(true);

    useEffect(() => {
        return () => {
            isMountedRef.current = false;
            cleanupFn();
        };
    }, [cleanupFn]);

    return isMountedRef;
};
