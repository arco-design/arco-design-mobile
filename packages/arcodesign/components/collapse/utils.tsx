import React, { Context } from 'react';
import { GroupContextParams } from './type';

export const defaultContextKey = '__COLLAPSE_CONTEXT__';
export const allContexts: Record<string, Context<GroupContextParams>> = {};

export function getCollapseContext(key: string) {
    if (!allContexts[key]) {
        allContexts[key] = React.createContext<GroupContextParams>({
            isGroup: false,
            value: [],
            onCollapse: () => {},
        });
    }
    return allContexts[key];
}

export const CollapseKeyContext = React.createContext<{ key: string }>({
    key: defaultContextKey,
});
