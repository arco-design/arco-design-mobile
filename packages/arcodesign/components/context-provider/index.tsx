import React, { createContext, useState, useEffect, useCallback } from 'react';
import { addCssRules, removeCssStyleDom, defaultLocale, ILocale } from '@arco-design/mobile-utils';

export interface GlobalContextParams {
    /**
     * 组件类名前缀
     * @en Component classname prefix
     * @default "arco"
     */
    prefixCls?: string;
    /**
     * 手动控制当前所在系统，传入后将直接使用传入的值，ssr场景需指定系统初始值时适用
     * @en Manually control the current system, and the incoming value will be used directly after being passed in. It is applicable when the initial value of the system needs to be specified in the ssr scenario.
     * @default ""
     */
    system?: 'pc' | 'android' | 'ios' | '';
    /**
     * 是否使用暗黑模式
     * @en Whether to use dark mode
     * @default false
     */
    useDarkMode?: boolean;
    /**
     * 是否处于暗黑模式，指定后以指定的值为准
     * @en Whether it is in dark mode
     * @default false
     */
    isDarkMode?: boolean;
    /**
     * 主题变量，传入后将在线替换css变量，需设置less变量 @use-css-vars: 1
     * @en Theme variable. The css variable will be replaced online after input. The less variable needs to be set @use-css-vars: 1
     */
    theme?: Record<string, string>;
    /**
     * 国际化语言包配置
     * @en Internationalized language configuration
     */
    locale?: ILocale;
    /**
     * 是否使用Rtl模式
     * @en Whether to use rtl
     * @default false
     */
    useRtl?: boolean;
}

export const defaultContext: GlobalContextParams = {
    prefixCls: 'arco',
    system: '',
    useDarkMode: false,
    isDarkMode: false,
    locale: defaultLocale,
    useRtl: false,
};

export const GlobalContext = createContext<GlobalContextParams>(defaultContext);

export interface ContextProviderProps extends GlobalContextParams {
    children: React.ReactNode;
}

export type WithGlobalContext<T> = T & { context?: GlobalContextParams };

/**
 * 全局数据控制组件，用于替换全局数据。
 * @en Global data control component, used to replace global data.
 * @type 其他
 * @type_en Others
 * @name 全局配置
 * @name_en ContextProvider
 */
export default function ContextProvider(props: ContextProviderProps) {
    const {
        children,
        useDarkMode,
        isDarkMode,
        theme,
        locale = defaultLocale,
        ...restProps
    } = props;
    const [isDarkModeState, setIsDarkModeState] = useState<boolean>();
    const changeDarkMode = useCallback(
        (res: MediaQueryListEvent) => setIsDarkModeState(res.matches),
        [],
    );

    useEffect(() => {
        const matchMedia = window.matchMedia('(prefers-color-scheme: dark)');
        if (useDarkMode) {
            const dark = matchMedia.matches;
            setIsDarkModeState(dark);
            if (typeof matchMedia.addEventListener === 'function') {
                matchMedia.addEventListener('change', changeDarkMode);
            } else if (typeof matchMedia.addListener === 'function') {
                matchMedia.addListener(changeDarkMode);
            }
        }
        return () => {
            if (useDarkMode) {
                if (typeof matchMedia.removeEventListener === 'function') {
                    matchMedia.removeEventListener('change', changeDarkMode);
                } else if (typeof matchMedia.removeListener === 'function') {
                    matchMedia.removeListener(changeDarkMode);
                }
            }
        };
    }, [useDarkMode]);

    useEffect(() => {
        if (!theme) {
            return;
        }
        addCssRules('arcoCustomTheme', theme);
        return () => {
            removeCssStyleDom('arcoCustomTheme');
        };
    }, [theme]);

    return (
        <GlobalContext.Provider
            value={{
                ...defaultContext,
                ...restProps,
                isDarkMode: isDarkMode === void 0 ? isDarkModeState : isDarkMode,
                useDarkMode,
                locale: locale || defaultLocale,
            }}
        >
            {children}
        </GlobalContext.Provider>
    );
}

export const ContextLayout = GlobalContext.Consumer;

export function CompWithGlobalContext<P extends JSX.IntrinsicAttributes>(
    Component: React.FunctionComponent<P>,
) {
    return function (props: WithGlobalContext<P>) {
        const { context: propsContext, ...others } = props;
        return (
            <ContextLayout>
                {context => (
                    <ContextProvider {...{ ...context, ...propsContext }}>
                        <Component {...(others as P)} />
                    </ContextProvider>
                )}
            </ContextLayout>
        );
    };
}
