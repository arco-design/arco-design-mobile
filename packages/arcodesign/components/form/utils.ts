import { IRules, ValidatorError } from '@arco-design/mobile-utils';

export const isFieldRequired = (rules: IRules[] = []) => {
    return (rules || []).some(rule => rule?.required);
};

export const getErrorAndWarnings = (result: ValidatorError[]) => {
    let errors: string[] = [];
    let warnings: string[] = [];
    result.map(({ message = [], validateLevel = 'error' }) => {
        if (!message?.length) {
            return;
        }
        if (validateLevel === 'warning') {
            warnings = [...warnings, ...message];
        } else {
            errors = [...errors, ...message];
        }
    });
    return { warnings, errors };
};
