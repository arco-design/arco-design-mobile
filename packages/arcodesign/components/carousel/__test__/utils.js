import { act } from 'react-dom/test-utils';
import { createMoveTouchEventObject, createStartTouchEventObject } from '../../../tests/helpers/mockEvent';

export function mockSwipeStart(map, clientXMap) {
    const { touchstart, touchmove } = clientXMap;
    act(() => {
        map.touchstart(createStartTouchEventObject({ x: touchstart, y: 0 }));
        map.touchmove(createMoveTouchEventObject({ x: touchmove, y: 0 }));
    });
}
export function mockSwipeEnd(wrapper, className, touchend) {
    wrapper.find(`.${className}`).simulate('touchEnd', createMoveTouchEventObject({ x: touchend, y: 0 }));
    act(() => {
        jest.advanceTimersByTime(600);
    });
    wrapper.update();
}

export function mockSwipe(map, wrapper, className, clientXMap) {
    const { touchend } = clientXMap;
    mockSwipeStart(map, clientXMap);
    mockSwipeEnd(wrapper, className, touchend);
}
