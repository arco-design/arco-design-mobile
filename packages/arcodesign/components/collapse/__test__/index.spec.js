import React from 'react';
import demoTest from '../../../tests/demoTest';
import mountTest from '../../../tests/mountTest';
import Collapse from '..';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { defaultContext } from '../../context-provider';
import { delay } from '../../../tests/helpers/utils';

const prefix = `${defaultContext.prefixCls}-collapse`;

demoTest('collapse');

mountTest(Collapse, 'Collapse');

const activeItem = [
    '1'
]
const item = [
    {
        header: 'title1',
        value: '1',
        content: 'content1'
    },
    {
        header: 'title2',
        value: '2',
        content: 'content2'
    },
    {
        header: 'title3',
        value: '3',
        content: 'content3'
    }
]

describe('Collapse', () => {

    beforeEach(() => {
        jest.useFakeTimers();
    });

    afterEach(() => {
        jest.useRealTimers();
    });

    it('Default expanded rendering is correct', () => {
        const wrapper = mount(
            <Collapse
                header="title1"
                value="1"
                content={
                    <div>content</div>
                } />,
        );
        wrapper.find(`.${prefix}-header`).at(0).simulate('click');
        expect(wrapper.find(`.${prefix}-content`).at(0).getDOMNode().style.height).toBe('auto');
    });

    it('Default Group expanded rendering is correct', () => {
        const mockFn = jest.fn();
        const wrapper = mount(
            <Collapse.Group
                items={item}
                activeItem={activeItem}
                onCollapse={mockFn}>
            </Collapse.Group>,
        );
        wrapper.setProps({
            activeItem: ['2']
        });
        wrapper.find(`.${prefix}-header`).at(0).simulate('click');
        expect(mockFn.mock.calls.length).toBe(1);
    });
    it('OnCollapse rendering is correct', () => {
        const mockFn = jest.fn();
        const wrapper = mount(
            <Collapse
                header="title1"
                value="1"
                content={
                    <div>content</div>
                }
                onCollapse={mockFn}/>,
        );
        wrapper.find(`.${prefix}-header`).at(0).simulate('click');
        expect(wrapper.find(`.${prefix}-content-container`).text()).toEqual('content');
        expect(mockFn.mock.calls.length).toBe(1);
    });
    it('Rerender is correct', () => {
        const wrapper = mount(
            <Collapse
                header="title1"
                value="1"
                content={
                    <div>content</div>
                }/>,
        );
        wrapper.setProps({
            content: (
                <div>
                    <div>content1</div>
                    <div>content1</div>
                    <div>content1</div>
                    <div>content1</div>
                </div>
            )
        })
        act(() => {
            jest.advanceTimersByTime(1000);
        });
        wrapper.update();
        wrapper.find(`.${prefix}-header`).at(0).simulate('click');
        expect(wrapper.find(`.${prefix}-content-container`).text()).toEqual('content1content1content1content1');
    });
    it('Disabled rendering is correct', () => {
        const wrapper = mount(
            <Collapse
                header="title3"
                value="3"
                disabled
                content={
                    <div>content</div>
                } />,
        );
        expect(wrapper.find(`.${prefix}-content`).at(0).getDOMNode().style.height).toBe('0px');
        wrapper.find(`.${prefix}-header`).simulate('click');
        expect(wrapper.find(`.${prefix}-content`).at(0).getDOMNode().style.height).toBe('0px');
    });
    it('Combined folds render correctly', () => {
        const wrapper = mount(
            <Collapse.Group defaultActiveItems={['1']} useAccordion>
                <Collapse
                    header="title1"
                    value="1"
                    content={
                        <Collapse.Group defaultActiveItems={['1']} useAccordion groupKey="sub-group">
                            <Collapse
                                header="children-title"
                                value="1"
                                content="content1"
                            />
                        </Collapse.Group>
                    }
                />
                <Collapse
                    header="content2"
                    active
                    value="2"
                    content="content2"
                />
            </Collapse.Group>,
        );
        expect(wrapper.find(`.${prefix}-content`).at(0).getDOMNode().style.height).toBe('auto');
        expect(wrapper.find(`.${prefix}-content`).at(1).getDOMNode().style.height).toBe('auto');
        expect(wrapper.find(`.${prefix}-content`).at(2).getDOMNode().style.height).toBe('0px');
        wrapper.find(`.${prefix}-header`).at(2).simulate('click');
        expect(wrapper.find(`.${prefix}-content`).at(0).getDOMNode().style.height).toBe('0px');
        expect(wrapper.find(`.${prefix}-content`).at(1).getDOMNode().style.height).toBe('auto');
        expect(wrapper.find(`.${prefix}-content`).at(2).getDOMNode().style.height).toBe('auto');
    });
})
