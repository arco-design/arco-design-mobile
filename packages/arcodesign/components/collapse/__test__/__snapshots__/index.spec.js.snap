// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collapse demo test collapse demo: children.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-collapse-group"
  >
    <div
      class="arco-collapse-item"
    >
      <div
        class="arco-collapse-header"
      >
        Title 1
        <div
          class="arco-collapse-icon"
        >
          <svg
            class="arco-icon arco-icon-arrow-down arco-collapse-icon-arrow"
            height="1em"
            style="transform: rotate(180deg);"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M512 676.864l346.944-346.965a21.333 21.333 0 0130.187 0l30.165 30.186a21.333 21.333 0 010 30.166L542.166 767.36a42.667 42.667 0 01-60.331 0L104.704 390.25a21.333 21.333 0 010-30.165l30.165-30.186a21.333 21.333 0 0130.187 0L512 676.864z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="arco-collapse-content"
        style="height: auto;"
      >
        <div
          class="arco-collapse-content-container"
        >
          <div
            class="arco-collapse-group"
          >
            <div
              class="arco-collapse-item"
            >
              <div
                class="arco-collapse-header"
              >
                Subtitle
                <div
                  class="arco-collapse-icon"
                >
                  <svg
                    class="arco-icon arco-icon-arrow-down arco-collapse-icon-arrow"
                    height="1em"
                    style="transform: rotate(180deg);"
                    viewBox="0 0 1024 1024"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M512 676.864l346.944-346.965a21.333 21.333 0 0130.187 0l30.165 30.186a21.333 21.333 0 010 30.166L542.166 767.36a42.667 42.667 0 01-60.331 0L104.704 390.25a21.333 21.333 0 010-30.165l30.165-30.186a21.333 21.333 0 0130.187 0L512 676.864z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="arco-collapse-content"
                style="height: auto;"
              >
                <div
                  class="arco-collapse-content-container"
                >
                  here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is the content area
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-collapse-item"
    >
      <div
        class="arco-collapse-header"
      >
        Title 2
        <div
          class="arco-collapse-icon"
        >
          <svg
            class="arco-icon arco-icon-arrow-down arco-collapse-icon-arrow"
            height="1em"
            style="transform: rotate(0deg);"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M512 676.864l346.944-346.965a21.333 21.333 0 0130.187 0l30.165 30.186a21.333 21.333 0 010 30.166L542.166 767.36a42.667 42.667 0 01-60.331 0L104.704 390.25a21.333 21.333 0 010-30.165l30.165-30.186a21.333 21.333 0 0130.187 0L512 676.864z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="arco-collapse-content"
        style="height: 0px;"
      >
        <div
          class="arco-collapse-content-container"
        >
          here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is the content area
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`collapse demo test collapse demo: group.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-collapse-group"
  >
    <div
      class="arco-collapse-item"
    >
      <div
        class="arco-collapse-header"
      >
        Title 1
        <div
          class="arco-collapse-icon"
        >
          <svg
            class="arco-icon arco-icon-arrow-down arco-collapse-icon-arrow"
            height="1em"
            style="transform: rotate(0deg);"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M512 676.864l346.944-346.965a21.333 21.333 0 0130.187 0l30.165 30.186a21.333 21.333 0 010 30.166L542.166 767.36a42.667 42.667 0 01-60.331 0L104.704 390.25a21.333 21.333 0 010-30.165l30.165-30.186a21.333 21.333 0 0130.187 0L512 676.864z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="arco-collapse-content"
        style="height: 0px;"
      >
        <div
          class="arco-collapse-content-container"
        >
          here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is the content area
        </div>
      </div>
    </div>
    <div
      class="arco-collapse-item"
    >
      <div
        class="arco-collapse-header"
      >
        Title 2
        <div
          class="arco-collapse-icon"
        >
          <svg
            class="arco-icon arco-icon-arrow-down arco-collapse-icon-arrow"
            height="1em"
            style="transform: rotate(0deg);"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M512 676.864l346.944-346.965a21.333 21.333 0 0130.187 0l30.165 30.186a21.333 21.333 0 010 30.166L542.166 767.36a42.667 42.667 0 01-60.331 0L104.704 390.25a21.333 21.333 0 010-30.165l30.165-30.186a21.333 21.333 0 0130.187 0L512 676.864z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="arco-collapse-content"
        style="height: 0px;"
      >
        <div
          class="arco-collapse-content-container"
        >
          here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is the content area
        </div>
      </div>
    </div>
    <div
      class="arco-collapse-item"
    >
      <div
        class="arco-collapse-header"
      >
        Title 3
        <div
          class="arco-collapse-icon"
        >
          <svg
            class="arco-icon arco-icon-arrow-down arco-collapse-icon-arrow"
            height="1em"
            style="transform: rotate(0deg);"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M512 676.864l346.944-346.965a21.333 21.333 0 0130.187 0l30.165 30.186a21.333 21.333 0 010 30.166L542.166 767.36a42.667 42.667 0 01-60.331 0L104.704 390.25a21.333 21.333 0 010-30.165l30.165-30.186a21.333 21.333 0 0130.187 0L512 676.864z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="arco-collapse-content"
        style="height: 0px;"
      >
        <div
          class="arco-collapse-content-container"
        >
          here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is the content area
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`collapse demo test collapse demo: index.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-collapse-item"
  >
    <div
      class="arco-collapse-header"
    >
      Title 1
      <div
        class="arco-collapse-icon"
      >
        <svg
          class="arco-icon arco-icon-arrow-down arco-collapse-icon-arrow"
          height="1em"
          style="transform: rotate(180deg);"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M512 676.864l346.944-346.965a21.333 21.333 0 0130.187 0l30.165 30.186a21.333 21.333 0 010 30.166L542.166 767.36a42.667 42.667 0 01-60.331 0L104.704 390.25a21.333 21.333 0 010-30.165l30.165-30.186a21.333 21.333 0 0130.187 0L512 676.864z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="arco-collapse-content"
      style="height: auto;"
    >
      <div
        class="arco-collapse-content-container"
      />
    </div>
  </div>
  <div
    class="arco-collapse-item"
  >
    <div
      class="arco-collapse-header"
    >
      Title 2
      <div
        class="arco-collapse-icon"
      >
        <svg
          class="arco-icon arco-icon-arrow-down arco-collapse-icon-arrow"
          height="1em"
          style="transform: rotate(0deg);"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M512 676.864l346.944-346.965a21.333 21.333 0 0130.187 0l30.165 30.186a21.333 21.333 0 010 30.166L542.166 767.36a42.667 42.667 0 01-60.331 0L104.704 390.25a21.333 21.333 0 010-30.165l30.165-30.186a21.333 21.333 0 0130.187 0L512 676.864z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="arco-collapse-content"
      style="height: 0px;"
    >
      <div
        class="arco-collapse-content-container"
      >
        here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is the content area
      </div>
    </div>
  </div>
  <div
    class="arco-collapse-item disabled"
  >
    <div
      class="arco-collapse-header"
    >
      Disabled
      <div
        class="arco-collapse-icon"
      >
        <svg
          class="arco-icon arco-icon-arrow-down arco-collapse-icon-arrow"
          height="1em"
          style="transform: rotate(0deg);"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M512 676.864l346.944-346.965a21.333 21.333 0 0130.187 0l30.165 30.186a21.333 21.333 0 010 30.166L542.166 767.36a42.667 42.667 0 01-60.331 0L104.704 390.25a21.333 21.333 0 010-30.165l30.165-30.186a21.333 21.333 0 0130.187 0L512 676.864z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="arco-collapse-content"
      style="height: 0px;"
    >
      <div
        class="arco-collapse-content-container"
      >
        here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is the content area
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`collapse demo test collapse demo: style-header.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-collapse-group"
  >
    <div
      class="arco-collapse-item"
    >
      <div
        class="arco-collapse-header"
      >
        <div
          style="display: flex; align-items: center;"
        >
          <svg
            class="arco-icon arco-icon-notice header-icon-notice"
            fill="currentColor"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs />
            <path
              d="M618.667 853.333A21.333 21.333 0 01640 874.667v42.666a21.333 21.333 0 01-21.333 21.334H405.333A21.333 21.333 0 01384 917.333v-42.666a21.333 21.333 0 0121.333-21.334h213.334zM576 64a21.333 21.333 0 0121.333 21.333v32.427c147.222 39.125 256 177.067 256 341.248V704h64a21.333 21.333 0 0121.334 21.333V768a21.333 21.333 0 01-21.334 21.333H106.667A21.333 21.333 0 0185.333 768v-42.667A21.333 21.333 0 01106.667 704h64V459.008c0-164.181 108.8-302.123 256-341.248V85.333A21.333 21.333 0 01448 64h128zm-64 128c-141.376 0-256 119.595-256 267.136V704h512V459.136C768 311.595 653.376 192 512 192z"
            />
          </svg>
          Title 1
        </div>
        <div
          class="arco-collapse-icon"
        >
          <svg
            class="arco-icon arco-icon-arrow-down arco-collapse-icon-arrow"
            height="1em"
            style="transform: rotate(0deg);"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M512 676.864l346.944-346.965a21.333 21.333 0 0130.187 0l30.165 30.186a21.333 21.333 0 010 30.166L542.166 767.36a42.667 42.667 0 01-60.331 0L104.704 390.25a21.333 21.333 0 010-30.165l30.165-30.186a21.333 21.333 0 0130.187 0L512 676.864z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="arco-collapse-content"
        style="height: 0px;"
      >
        <div
          class="arco-collapse-content-container"
        >
          here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is the content area
        </div>
      </div>
    </div>
    <div
      class="arco-collapse-item"
    >
      <div
        class="arco-collapse-header"
      >
        <div
          style="display: flex; align-items: center;"
        >
          Title 2
          <svg
            class="arco-icon arco-icon-question-circle header-icon-question-circle"
            fill="currentColor"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M7.982.68a7.312 7.312 0 110 14.625 7.312 7.312 0 010-14.625zm0 1.33a5.983 5.983 0 100 11.965 5.983 5.983 0 000-11.966zm.333 8.641c.183 0 .332.15.332.333v.664a.332.332 0 01-.332.333H7.65a.332.332 0 01-.332-.333v-.664c0-.184.148-.333.332-.333h.665zm-.427-6.647c.79 0 1.447.19 1.97.574.522.382.784.95.784 1.7 0 .461-.114.85-.341 1.165-.133.191-.389.436-.766.733l-.373.293c-.202.16-.337.345-.403.558a1.8 1.8 0 00-.048.323.332.332 0 01-.331.304h-.744a.332.332 0 01-.331-.363c.043-.463.088-.755.137-.878.091-.229.325-.492.703-.79l.383-.302c.126-.096.774-.553.774-.957 0-.405-.07-.553-.273-.775-.203-.222-.66-.294-1.073-.294-.405 0-.752.108-.922.38a1.921 1.921 0 00-.189.4 1.388 1.388 0 00-.048.204.332.332 0 01-.328.276h-.782a.332.332 0 01-.33-.37c.008-.074.017-.136.025-.185.13-.722.458-1.253.985-1.592.413-.27.92-.404 1.521-.404z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="arco-collapse-icon"
        >
          <svg
            class="arco-icon arco-icon-arrow-down arco-collapse-icon-arrow"
            height="1em"
            style="transform: rotate(0deg);"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M512 676.864l346.944-346.965a21.333 21.333 0 0130.187 0l30.165 30.186a21.333 21.333 0 010 30.166L542.166 767.36a42.667 42.667 0 01-60.331 0L104.704 390.25a21.333 21.333 0 010-30.165l30.165-30.186a21.333 21.333 0 0130.187 0L512 676.864z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="arco-collapse-content"
        style="height: 0px;"
      >
        <div
          class="arco-collapse-content-container"
        >
          here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is content area, here is the content area
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
